src/utils/utils.ts

```diff
@@ -282,3 +297,122 @@ export const groupByFirstLevelItems = (
   );
   return d;
 };
+
+const getStartDateFunction = (interval: GroupByInterval) => {
+  switch (interval) {
+    case GroupByInterval.Year:
+      return startOfYear;
+    case GroupByInterval.Month:
+      return startOfMonth;
+    case GroupByInterval.Day:
+      return startOfDay;
+    case GroupByInterval.Week:
+    default:
+      return (date: Date) => startOfWeek(date, { weekStartsOn: 1 });
+  }
+};
+
+const fillDateGaps = (
+  data: { [key: string]: Action[] },
+  freq: GroupByInterval,
+  start: Date,
+  stop: Date,
+): { [key: string]: Action[] } => {
+  const copy = { ...data };
+  const nextFunc = {
+    [GroupByInterval.Day]: addDays,
+    [GroupByInterval.Month]: addMonths,
+    [GroupByInterval.Week]: addWeeks,
+    [GroupByInterval.Year]: addYears,
+  };
+  const getStartDate = getStartDateFunction(freq);
+
+  let currentDate = getStartDate(start);
+
+  while (isBefore(currentDate, stop)) {
+    if (!copy[currentDate.toISOString()]) {
+      copy[currentDate.toISOString()] = [];
+    }
+
+    const getNextInterval = nextFunc[freq];
+    currentDate = getNextInterval(currentDate, 1);
+  }
+
+  return copy;
+};
+
+const groupActionsByInterval = (
+  actions: Action[],
+  interval: GroupByInterval,
+): { [key: string]: Action[] } => {
+  const getStartDate = getStartDateFunction(interval);
+
+  const groupedByInterval = groupBy(actions, (item) => {
+    const date = parseISO(item.createdAt);
+    const startDate = getStartDate(date);
+    return startDate.toISOString();
+  });
+
+  return groupedByInterval;
+};
+
+const groupActionsBasedOnMaxIntervals = (
+  actions: {
+    [key: string]: Action[];
+  },
+  maxNoOfIntervals: number,
+): { [key: string]: Action[] } => {
+  const keys = Object.keys(actions);
+  const totalActions = Object.values(actions);
+
+  const len = keys.length;
+
+  const maxKeysPerGroup = Math.ceil(len / maxNoOfIntervals);
+
+  if (len > maxNoOfIntervals) {
+    const obj: { [key: string]: Action[] } = {};
+    for (let index = 0; index < len; index += maxKeysPerGroup) {
+      const actionsOverInterval = totalActions.slice(
+        index,
+        index + maxKeysPerGroup,
+      );
+
+      const totals = actionsOverInterval.reduce((acc, curr) => {
+        return [...acc, ...curr];
+      }, []);
+
+      obj[keys[index]] = totals;
+    }
+    return obj;
+  }
```

Comment:
I think all this could have been a big `reduce` function? Maybe based on `idx` of `totalActions`, you can get use 
```
acc[idx % maxNoOfIntervals] = acc[idx % maxNoOfIntervals].concat(curr)
```
where initialized value is `Array.from({length:maxKeysPerGroup},()=>[])`
If I'm not wrong